File: src/utils/llm_wrapper.py
from openai import OpenAI
from openai.types.chat import ChatCompletion
from pydantic import BaseModel

from .config import read_config


def chat(
    system_prompt: str,
    user_prompt: str,
    response_model: BaseModel,
) -> BaseModel:
    provider_params, request_params = _get_params()

    client: OpenAI = OpenAI(**provider_params)
    response: ChatCompletion = client.chat.completions.parse(
        **request_params,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        max_tokens=1024,
        temperature=0,
        response_format=response_model,
    )
    return response.choices[0].message.parsed


def _get_params() -> tuple[dict, dict]:
    config: dict = read_config()
    model: str = config.get("MODEL")
    if not model:
        raise ValueError("MODEL not found in config")

    provider, model_name = model.lower().split(":", 1)

    provider_params: dict = {}
    request_params: dict = {}

    if provider == "openai":
        api_key: str = config.get("OPENAI_API_KEY")
        if not api_key:
            raise ValueError("OPENAI_API_KEY not found in environment")
        provider_params["api_key"] = api_key
    elif provider == "anthropic":
        api_key: str = config.get("ANTHROPIC_API_KEY")
        if not api_key:
            raise ValueError("ANTHROPIC_API_KEY not found in environment")
        provider_params["api_key"] = api_key
        provider_params["base_url"] = "https://api.anthropic.com/v1/"
    elif provider == "ollama":
        api_key: str = config.get("GOOGLE_API_KEY")
        if not api_key:
            raise ValueError("GOOGLE_API_KEY not found in environment")
        provider_params["api_key"] = api_key
        provider_params["base_url"] = (
            "https://generativelanguage.googleapis.com/v1beta/openai/"
        )
    else:
        raise ValueError(f"Invalid provider: {provider}")

    reasoning_models: list[str] = [
        "gemini-2.5-flash-lite",
        "gemini-2.5-flash",
        "gemini-2.5-pro",
    ]
    if model_name in reasoning_models:
        request_params["reasoning_effort"] = "none"

    request_params["model"] = model_name

    return provider_params, request_params

File: src/utils/llm-wrapper.ts
import { createAnthropic } from "@ai-sdk/anthropic";
import { createGoogleGenerativeAI } from "@ai-sdk/google";
import { createOpenAI } from "@ai-sdk/openai";
import { generateObject } from "ai";
import type { z } from "zod";
import { type Config, readConfig } from "./config";

export interface ModelParams {
  provider: string;
  modelName: string;
  apiKey: string;
}

export function getModelParams(): ModelParams {
  const config: Config = readConfig();

  if (!config.MODEL) {
    throw new Error("MODEL not found in config");
  }

  const [provider, modelName] = config.MODEL.toLowerCase().split(":", 2);

  if (!provider || !modelName) {
    throw new Error('Invalid MODEL format. Expected "provider:model-name"');
  }

  let apiKey: string | undefined;

  switch (provider) {
    case "openai":
      apiKey = config.OPENAI_API_KEY;
      if (!apiKey) {
        throw new Error("OPENAI_API_KEY not found in config");
      }
      break;
    case "anthropic":
      apiKey = config.ANTHROPIC_API_KEY;
      if (!apiKey) {
        throw new Error("ANTHROPIC_API_KEY not found in config");
      }
      break;
    case "google":
      apiKey = config.GOOGLE_API_KEY;
      if (!apiKey) {
        throw new Error("GOOGLE_API_KEY not found in config");
      }
      break;
    default:
      throw new Error(`Invalid provider: ${provider}`);
  }

  return { provider, modelName, apiKey };
}

export function getModel(params: ModelParams) {
  const { provider, modelName, apiKey } = params;

  switch (provider) {
    case "openai": {
      const openai = createOpenAI({ apiKey });
      return openai(modelName);
    }
    case "anthropic": {
      const anthropic = createAnthropic({ apiKey });
      return anthropic(modelName);
    }
    case "google": {
      const google = createGoogleGenerativeAI({ apiKey });
      return google(modelName);
    }
    default:
      throw new Error(`Unsupported provider: ${provider}`);
  }
}

export async function chat<T extends z.ZodType>(
  systemPrompt: string,
  userPrompt: string,
  schema: T,
): Promise<z.infer<T>> {
  const params = getModelParams();
  const model = getModel(params);

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 60000);

  try {
    const result = await generateObject({
      model,
      schema,
      system: systemPrompt,
      prompt: userPrompt,
      abortSignal: controller.signal,
      providerOptions: {
        openai: {
          reasoningEffort: "minimal",
        },
        anthropic: {
          reasoningEffort: "minimal",
        },
        google: {
          reasoningEffort: "minimal",
        },
      },
    });

    return result.object as z.infer<T>;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    throw new Error(
      `Failed to generate object using ${params.provider}:${params.modelName}: ${errorMessage}`,
    );
  } finally {
    clearTimeout(timeoutId);
  }
}
